File: launch_school/Programming_Foundations/small_problems/leading_substrings.rb
def leading_substrings(str)
  str.size <= 1 ? [str] : (leading_substrings(str[..-2]) + [str])
end

def substrings(str)
  str.size <= 1 ? [str] : (leading_substrings(str) + substrings(str[1..]))
end

# p leading_substrings('abc') == ['a', 'ab', 'abc']
# p leading_substrings('') #== ['a']
# p leading_substrings('xyzzy') == ['x', 'xy', 'xyz', 'xyzz', 'xyzzy']

p substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

File: launch_school/LS215-Computational-Thinking/small_problems/list_processing/leading_substrings.js
function leadingSubstrings(string) {
  let collection = [];

  [...string].reduce((acc, char) => {
    collection.push(acc);
    return acc + char;
  });

  collection.push(string);

  console.log(collection);
  return collection;
}

leadingSubstrings('abc');      // ["a", "ab", "abc"]
leadingSubstrings('a');        // ["a"]
leadingSubstrings('xyzzy');    // ["x", "xy", "xyz", "xyzz", "xyzzy"]
